swarm/storage/netstore.go

```diff
@@ -18,318 +18,280 @@ package storage
 		}
-		// The chunk is not available in the LocalStore, let's get the fetcher for it, or create a new one
-		// if it doesn't exist yet
-		f := n.getOrCreateFetcher(ctx, ref)
-		// If the caller needs the chunk, it has to use the returned fetch function to get it
-		return nil, f.Fetch, nil
-	}
 
-	return chunk, nil, nil
-}
+		if req.HopCount >= maxHopCount {
+			return nil, fmt.Errorf("reach %v hop counts for ref=%s", maxHopCount, fmt.Sprintf("%x", req.HopCount))
+		}
 
-// getOrCreateFetcher attempts at retrieving an existing fetchers
-// if none exists, creates one and saves it in the fetchers cache
-// caller must hold the lock
-func (n *NetStore) getOrCreateFetcher(ctx context.Context, ref Address) *fetcher {
-	if f := n.getFetcher(ref); f != nil {
-		return f
-	}
+		log.Trace("netstore.chunk-not-in-localstore", "ref", ref.String(), "hopCount", req.HopCount)
+		v, err, _ := n.requestGroup.Do(ref.String(), func() (interface{}, error) {
+			// currently we issue a retrieve request if a fetcher
+			// has already been created by a syncer for that particular chunk.
+			// so it is possible to
+			// have 2 in-flight requests for the same chunk - one by a
+			// syncer (offered/wanted/deliver flow) and one from
+			// here - retrieve request
+			has, fi, _ := n.HasWithCallback(ctx, ref, "request")
+			if !has {
+				err := n.RemoteFetch(ctx, req, fi)
+				if err != nil {
+					return nil, err
+				}
+			}
 
-	// no fetcher for the given address, we have to create a new one
-	key := hex.EncodeToString(ref)
-	// create the context during which fetching is kept alive
-	cctx, cancel := context.WithTimeout(ctx, fetcherTimeout)
-	// destroy is called when all requests finish
-	destroy := func() {
-		// remove fetcher from fetchers
-		n.fetchers.Remove(key)
-		// stop fetcher by cancelling context called when
-		// all requests cancelled/timedout or chunk is delivered
-		cancel()
-	}
-	// peers always stores all the peers which have an active request for the chunk. It is shared
-	// between fetcher and the NewFetchFunc function. It is needed by the NewFetchFunc because
-	// the peers which requested the chunk should not be requested to deliver it.
-	peers := &sync.Map{}
+			ch, err := n.Store.Get(ctx, mode, ref)
+			if err != nil {
+				log.Error(err.Error(), "ref", ref)
+				return nil, errors.New("item should have been in localstore, but it is not")
+			}
 
-	cctx, sp := spancontext.StartSpan(
-		cctx,
-		"netstore.fetcher",
-	)
+			// fi could be nil if the chunk was added to the NetStore between n.store.Get and the call to n.HasWithCallback
+			if fi != nil {
+				metrics.GetOrRegisterResettingTimer(fmt.Sprintf("fetcher.%s.request", fi.CreatedBy), nil).UpdateSince(start)
+			}
 
-	sp.LogFields(olog.String("ref", ref.String()))
-	fetcher := newFetcher(sp, ref, n.NewNetFetcherFunc(cctx, ref, peers), destroy, peers, n.closeC)
-	n.fetchers.Add(key, fetcher)
+			return ch, nil
+		})
 
-	return fetcher
-}
+		res, _ := v.(Chunk)
```

Comment:
Because the code is simple and it is clear that in all cases we return a Chunk. But you're right, I will probably add a panic here, because if we fuck this up, I'd rather we have Swarm panic right away.

Reply:
removed verbosity as Viktor suggested.